### PR TITLE
Add terraform module

### PR DESCRIPTION
**Pull request by Abuelodelanada**
This PR adds a terraform module for the mimir-coordinator charm, that can be used in tandem with the [terraform module for the mimir-worker](https://github.com/canonical/mimir-worker-k8s-operator/pull/74) charm

## Testing instructions

1. Create a juju model, for instance "mimir"
2. Copy the following content in a file named mimir.tf
3. Run: `terraform init` and `terraform apply`

```terraform
terraform {
  required_version = ">= 1.5"
  required_providers {
    juju = {
      source  = "juju/juju"
      version = "~> 0.14.0"
    }
  }
}

provider "juju" {}

variable "model_name" {
  description = "Model name"
  type        = string
}


module "read" {
  source = "git::https://github.com/canonical/mimir-worker-k8s-operator.git//terraform?ref=feature/terraform"
  channel = "latest/edge"
  units = 3
  app_name = "worker-read"
  model_name = var.model_name
  config = {"role-read": "true"}
}

module "write" {
  source = "git::https://github.com/canonical/mimir-worker-k8s-operator.git//terraform?ref=feature/terraform"
  channel = "latest/edge"
  units = 3
  app_name = "worker-write"
  model_name = var.model_name
  config = {"role-write": "true"}
}
module "backend" {
  source = "git::https://github.com/canonical/mimir-worker-k8s-operator.git//terraform?ref=feature/terraform"
  channel = "latest/edge"
  units = 3
  app_name = "worker-backend"
  model_name = var.model_name
  config = {"role-backend": "true"}
}
module "coord" {
  source = "git::https://github.com/canonical/mimir-coordinator-k8s-operator.git//terraform?ref=feature/terraform"
  app_name = "mimir"
  model_name = var.model_name
}


resource "juju_application" "minio" {
  name = "minio"
  # Coordinator requires s3
  model = var.model_name
  trust = true

  charm {
    name    = "minio"
    channel = "latest/edge"
  }
  units = 1

  config = {
    access-key = "user"
    secret-key = "password"
  }
}

resource "juju_application" "s3int" {
  name = "s3int"

  model = var.model_name
  trust = true

  charm {
    name    = "s3-integrator"
    channel = "latest/edge"
  }
  units = 1

  config = {
    endpoint = "${juju_application.minio.name}-0.${juju_application.minio.name}-endpoints.${var.model_name}.svc.cluster.local:9000"
    bucket   = "mimir"
  }
}


resource "null_resource" "s3fix" {

  provisioner "local-exec" {
    # There's currently no way to wait for the charm to be idle, hence the sleep
    # https://github.com/juju/terraform-provider-juju/issues/202
    command = "sleep 120;juju run -m ${var.model_name} s3int/leader sync-s3-credentials access-key=user secret-key=password;juju ssh -m ${var.model_name} minio/leader curl https://dl.min.io/client/mc/release/linux-amd64/mc --create-dirs -o '/root/minio/mc';juju ssh -m ${var.model_name} minio/leader chmod +x '/root/minio/mc';juju ssh -m ${var.model_name} minio/leader /root/minio/mc mb mimir;sleep 30"
  }
}

resource "juju_integration" "coordinator_to_s3int" {
  model = var.model_name

  application {
    name     = juju_application.s3int.name
    endpoint = "s3-credentials"
  }

  application {
    name     = module.coord.app_name
    endpoint = "s3"
  }
}


# Relations
resource "juju_integration" "read_to_coordinator" {
  model = var.model_name

  application {
    name     = module.coord.app_name
    endpoint = "mimir-cluster"
  }

  application {
    name     = module.read.app_name
    endpoint = "mimir-cluster"
  }
}

resource "juju_integration" "write_to_coordinator" {
  model = var.model_name

  application {
    name     = module.coord.app_name
    endpoint = "mimir-cluster"
  }

  application {
    name     = module.write.app_name
    endpoint = "mimir-cluster"
  }
}

resource "juju_integration" "backend_to_coordinator" {
  model = var.model_name

  application {
    name     = module.coord.app_name
    endpoint = "mimir-cluster"
  }

  application {
    name     = module.backend.app_name
    endpoint = "mimir-cluster"
  }
}
```

### CODE REVIEW

**Review by sed-i — APPROVED** [review]
Tested indirectly by deploying https://github.com/canonical/observability/pull/212.

File: .gitignore [hunk 1]
```diff
@@ -8,4 +8,5 @@ __pycache__/
 .idea
 .vscode/
 *.egg-info/
-cos-tool-*
\ No newline at end of file
+cos-tool-*
+*.terraform*
```
Comment (MichaelThamm) [inline .gitignore:12]: Not sure what we should standardise our .gitignores to for TF but I found this from one of our spec examples:

https://github.com/canonical/grafana-k8s-operator/pull/350/files#diff-bc37d034bad564583790a46f19d807abfe519c5671395fd494d8cce506c42947

File: terraform/versions.tf [hunk 2]
```diff
@@ -0,0 +1,9 @@
+terraform {
+  required_version = ">= 1.5"
+  required_providers {
+    juju = {
+      source  = "juju/juju"
+      version = "~> 0.14.0"
```
Comment (MichaelThamm) [inline terraform/versions.tf:6]: It is probably a good idea if we do not lock our TF deployments to version 0.14:

[Look at this](https://github.com/canonical/observability/pull/212/files#diff-f050dd035c535a9dc5d61f4514cb38df2338426605c532ef89573aaf77f14195)

**Comment by MichaelThamm** [comment]
Overall, looks great! I just left 2 comments for you to consider.

**Pull request merged**